oxide-auth-async/src/tests/access_token.rs

```diff
@@ -0,0 +1,669 @@
+use oxide_auth::primitives::authorizer::AuthMap;
+use oxide_auth::primitives::issuer::TokenMap;
+use oxide_auth::primitives::grant::{Grant, Extensions};
+use oxide_auth::{
+    frontends::simple::endpoint::Error,
+    primitives::registrar::{Client, ClientMap},
+    endpoint::WebRequest,
+};
+
+use crate::{
+    endpoint::{access_token::AccessTokenFlow, Endpoint},
+    primitives::Authorizer,
+    frontends::simple::endpoint::{OptAuthorizer, OptRegistrar, OptIssuer, Vacant, ResponseCreator},
+};
+//use crate::frontends::simple::endpoint::access_token_flow;
+
+use std::collections::HashMap;
+
+use base64;
+use chrono::{Utc, Duration};
+use serde_json;
+
+use super::{Body, CraftedRequest, CraftedResponse, Status, TestGenerator, ToSingleValueQuery};
+use super::defaults::*;
+
+struct AccessTokenSetup {
+    registrar: ClientMap,
+    authorizer: AuthMap<TestGenerator>,
+    issuer: TokenMap<TestGenerator>,
+    authtoken: String,
+    basic_authorization: String,
+}
+
+struct AccessTokenEndpoint<'a> {
+    registrar: &'a ClientMap,
+    authorizer: &'a mut AuthMap<TestGenerator>,
+    issuer: &'a mut TokenMap<TestGenerator>,
+    response: Vacant,
+}
+
+impl<'a> AccessTokenEndpoint<'a> {
+    pub fn new(
+        registrar: &'a ClientMap, authorizer: &'a mut AuthMap<TestGenerator>,
+        issuer: &'a mut TokenMap<TestGenerator>,
+    ) -> Self {
+        AccessTokenEndpoint {
+            registrar,
+            authorizer,
+            issuer,
+            response: Vacant,
+        }
+    }
+}
+
+impl<'a> Endpoint<CraftedRequest> for AccessTokenEndpoint<'a> {
```

Comment:
I do have one question not directly related to the specific PR here. Did having the define this `impl` feel more cumbersome than using the `simple::Generic` endpoint in the synchronous version? The reason I'm asking is that, in the past several different pieces of feedback hinted that it was hard to discover that you do not _need_ to use `Generic` and that defining a custom trait impl can be much more ergonomic in complex cases. Therefore I was considering moving away from providing `simple::Generic` in the first place, or solely moving it into the tests. I'm just not sure if this removes an interface with usability advantages compared to the 'manual' impl.

Reply:
Indeed, when using the lib for the first time, I was quite confuse on if I should use `Generic` or not. At the moment, I'm using it in my app, but now that I've done this Endpoint myself, it seems easy enough that I'm considering to move away from Generic :) 

Dealing with generic type is quite hard to track, at least, it is for my brain :D